masq/src/commands/financials_command.rs

```diff
@@ -58,26 +175,911 @@ impl Command for FinancialsCommand {
+    }
+}
+
+#[derive(Debug, PartialEq)]
+enum ChangeDone<T> {
+    Changed(T),
+    Unchanged(T),
+}
+
+impl<T> ChangeDone<T> {
+    fn value(self) -> T {
+        match self {
+            Changed(val) => val,
+            Unchanged(val) => val,
+        }
+    }
+}
+
+trait Headings {
+    fn words(&self) -> &[&'static str];
+    fn is_gwei(&self) -> bool;
+}
+
+struct PayableHeadings(Vec<&'static str>, bool);
+impl Headings for PayableHeadings {
+    fn words(&self) -> &[&'static str] {
+        &self.0
+    }
+
+    fn is_gwei(&self) -> bool {
+        self.1
+    }
+}
+
+struct ReceivableHeadings(Vec<&'static str>, bool);
+impl Headings for ReceivableHeadings {
+    fn words(&self) -> &[&'static str] {
+        &self.0
+    }
+
+    fn is_gwei(&self) -> bool {
+        self.1
+    }
+}
+
+trait WidthInfo {
+    fn widths(&self, is_gwei: bool) -> Vec<usize>;
+}
+
+//we can ignore a special case of width computation of balances in MASQ with dot and two decimal digits,
+//it would require more MASQ than exist to grow wider beyond the heading of the column
+impl WidthInfo for UiPayableAccount {
+    fn widths(&self, is_gwei: bool) -> Vec<usize> {
+        vec![
+            //TODO can be computed as the whole number of accounts and doing this on it, just once
+            FinancialsCommand::count_length_with_comma_separators(self.age, true),
+            FinancialsCommand::count_length_with_comma_separators(self.balance_gwei, is_gwei),
+            if let Some(transaction_hash) = &self.pending_payable_hash_opt {
+                transaction_hash.len()
+            } else {
+                0
+            },
+        ]
+    }
+}
+
+impl WidthInfo for UiReceivableAccount {
+    fn widths(&self, is_gwei: bool) -> Vec<usize> {
+        vec![
+            FinancialsCommand::count_length_with_comma_separators(self.age, true),
+            FinancialsCommand::count_length_with_comma_separators(self.balance_gwei, is_gwei),
+        ]
+    }
+}
+
+pub fn validate_two_ranges<N>(double: String) -> Result<(), String>
+where
+    N: FromStr<Err = ParseIntError>
+        + From<u32>
+        + Mul<Output = N>
+        + PartialOrd
+        + Display
+        + CheckedMul
+        + MaxValue,
+{
+    fn checked_division<'a>(
+        str: &'a str,
+        delim: char,
+        err_msg: &'a str,
+    ) -> Result<Vec<&'a str>, String> {
+        let ranges = str.split(delim).collect::<Vec<&str>>();
+        if ranges.len() != 2 {
+            return Err(err_msg.to_string());
+        }
+        Ok(ranges)
+    }
+    let separate_ranges = checked_division(&double, '|', "Central vertical delimiter misused")?;
+    let time_range = checked_division(separate_ranges[0], '-', "First range is formatted wrong")?;
+    let (min_age, max_age) = parse_time_params(&time_range)?;
+    let (min_amount, max_amount, _, _): (N, N, _, _) =
+        parse_masq_range_to_gwei(separate_ranges[1])?;
+    if min_age >= max_age || min_amount >= max_amount {
```

Comment:
Maybe equality shouldn't be rejected?

Reply:
I was thinking a lot about this and my opinion in the end is that there are technical aspects that would take this away from the user.
a) You could probably never guess the exact second...as it's an all time moving value.
b) We could argue that exact query on balance makes sense. Well, I thought that too, but then I realized that we mask every query to whole billions, because we don't care about values smaller than 1 Gwei. At some point, I chose to simplify also the transfer of parameters chosen by the user, so he cannot specify anything smaller than 1 Gwei to be queried. He inputs Gweis. And so if he have only whole Gweis and the database stores even Weis, he is overplayed with the dust, that never allows to match the numbers.
Unless I introduce a rounding function in the sql... I'll check it out.